pkg/kv/iterators.go

```diff
@@ -0,0 +1,275 @@
+// the proto message and primary key
+type SecondaryIterator struct {
+	ctx          context.Context
+	itr          PrimaryIterator
+	partitionKey string
+	store        Store
+	msgType      protoreflect.MessageType
+	value        *MessageEntry
+	err          error
+}
+
+func NewSecondaryIterator(ctx context.Context, store Store, msgType protoreflect.MessageType, partitionKey string, prefix, after []byte) (*SecondaryIterator, error) {
+	itr, err := NewPrimaryIterator(ctx, store, (&SecondaryIndex{}).ProtoReflect().Type(), partitionKey, prefix, after)
+	if err != nil {
+		return nil, fmt.Errorf("create prefix iterator: %w", err)
+	}
+	return &SecondaryIterator{ctx: ctx, itr: *itr, partitionKey: partitionKey, store: store, msgType: msgType}, nil
+}
+
+func (s *SecondaryIterator) Next() bool {
+	if s.Err() != nil {
+		return false
+	}
+	if !s.itr.Next() {
+		return false
+	}
+	secondary := s.itr.Entry()
+	if secondary == nil {
+		s.err = ErrNotFound
+		return false
+	}
+	next := secondary.Value.(*SecondaryIndex)
+
+	var (
+		primary *ValueWithPredicate
+		err     error
+	)
+	for {
+		primary, err = s.store.Get(s.ctx, []byte(s.partitionKey), next.PrimaryKey)
+		if !errors.Is(err, ErrNotFound) {
+			break
+		}
+		if !s.itr.Next() {
+			return false
+		}
+		secondary = s.itr.Entry()
+		if secondary == nil {
+			s.err = ErrNotFound
+			return false
+		}
+		next = secondary.Value.(*SecondaryIndex)
+	}
+	if err != nil {
+		s.err = fmt.Errorf("getting value from key (primary key %s): %w", next.PrimaryKey, err)
+		return false
+	}
+	value := s.msgType.New().Interface()
+	err = proto.Unmarshal(primary.Value, value)
+	if err != nil {
+		s.err = fmt.Errorf("unmarshal proto data for key %s: %w", next.PrimaryKey, err)
+		return false
+	}
+	s.value = &MessageEntry{
+		Key:   string(next.PrimaryKey),
+		Value: value,
+	}
+	return true
+}
+
+func (s *SecondaryIterator) Entry() *MessageEntry {
+	return s.value
+}
+
+func (s *SecondaryIterator) Err() error {
+	if s.err != nil {
+		return s.err
+	}
+	return s.itr.Err()
+}
+
+func (s *SecondaryIterator) Close() {
+	s.itr.Close()
+}
+
+// SkipFirstIterator will keep the behaviour of the given EntriesIterator,
+// except for skipping the first Entry if its Key is equal to 'after'.
+type SkipFirstIterator struct {
+	it         EntriesIterator
+	after      []byte
+	nextCalled bool
+}
+
+func NewSkipIterator(it EntriesIterator, after []byte) EntriesIterator {
+	return &SkipFirstIterator{it: it, after: after}
+}
+
+func (si *SkipFirstIterator) Next() bool {
+	if !si.nextCalled {
+		si.nextCalled = true
+		if !si.it.Next() {
+			return false
+		}
+		if !bytes.Equal(si.it.Entry().Key, si.after) {
+			return true
+		}
+	}
+	return si.it.Next()
+}
+
+func (si *SkipFirstIterator) Entry() *Entry {
+	return si.it.Entry()
+}
+
+func (si *SkipFirstIterator) Err() error {
+	return si.it.Err()
+}
+
+func (si *SkipFirstIterator) Close() {
+	si.it.Close()
+}
+
+type PartitionIterator struct {
```

Comment:
I think we're missing tests for this iterator.

Reply:
We are missing unit tests for all kv iterators - let's open a new issue for that